### PR TITLE
add batch-combine workflows for Terra before MergeCohortVcfs and 05/06

### PR DESCRIPTION
**Pull request by epiercehoffman**
WDLs for workaround to create cohort-level sample set in Terra with the necessary aggregated files from individual batches. Ideally this would be done with a sample superset/set of sample sets data entity type, but such an entity type does not exist in Terra. It also currently seems to not be possible to upload arrays of files in a sample sets TSV, which limited the approach to the workaround. 

These workflows are designed to fit into the pipeline as follows:

1. Run up through Module03 on samples/sets as appropriate
2. Download the sample sets table as a TSV and upload it to Google Cloud
3. Run BatchCombinePreMerge.wdl with the sample sets TSV URI as input (and the batches to combine, if not all) on a new cohort-level sample set 
4. Run MergeCohortVcfs.wdl and output to Workspace Data
5. Run Modules 04 and 04b on each batch
6. Download the updated sample sets TSV and upload it to Google Cloud
7. Run BatchCombinePre0506.wdl on the existing cohort-level sample set, with the new sample sets TSV URI 
8. Run Modules 05_06 and 07 on the cohort-level sample set

Both BatchCombine workflows use the SelectBatchesAndColumns task, which parses the sample sets TSV to select one column at a time for the specified batches (or all batches if none specified).

Code temporarily hosted on a personal repo (https://github.com/epiercehoffman/gatk-sv-terra-eph) in order to perform testing in the following Terra workspaces (slack DM to be added):
* https://app.terra.bio/#workspaces/fccredits-barium-pear-9593/mod00a%20copy/data
* https://app.terra.bio/#workspaces/fccredits-barium-pear-9593/gatk-sv-pipeline%20copy/data

### CODE REVIEW

File: wdl/SelectBatchesAndColumns.wdl [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	command <<<
+		set -euo pipefail
+		1>&2 echo "Selecting columns from sample sets table"
+		FILE=~{sample_sets_table_tsv}
+
+		COLNAME=~{column}
+		1>&2 echo "COLUMN NAME=$COLNAME"
+
+		COLNUM=$(head -1 $FILE | awk -v RS='\t' -v field="$COLNAME" '$0~field{print NR; exit}')
+		1>&2 echo "COLUMN NUMBER=$COLNUM"
```
Comment (mwalker174) [inline wdl/SelectBatchesAndColumns.wdl:29]: What happens if the column isn't found? We would want it to catch and exit gracefully with an error message.

File: wdl/SelectBatchesAndColumns.wdl [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+		# if batches specified, select rows containing given batches in 1st column then select specified column
+		# otherwise, select specified column from all rows
+		# delete quotes and brackets from any bracketed, comma-separated arrays and split by newlines
+		for BATCH in '~{sep="' '" batches}'; do
+			1>&2 echo "BATCH=$BATCH"
+			~{if select_batches then "1>&2 echo 'Selecting batches from sample sets table'; awk -v FS='\t' -v bat=$BATCH '$1 == bat' $FILE" else "cat $FILE"} \
```
Comment (mwalker174) [inline wdl/SelectBatchesAndColumns.wdl:36]: Also here, what happens if a specified batch isn't found?

File: wdl/BatchCombinePre0506.wdl [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import "SelectBatchesAndColumns.wdl" as select
+
+workflow BatchCombinePre0506 {
+	input {
+		File sample_sets_table_tsv
+		Array[String]? batches
+		String? docker
+	}
+	Array[String] columns=["trained_genotype_depth_depth_sepcutoff","regenotyped_depth_vcfs", "merged_bincov", "merged_PE","median_cov","ped_file_postOutlierExclusion", "genotyped_pesr_vcf", "sr_background_fail","sr_bothside_pass","cutoffs"]
```
Comment (mwalker174) [inline wdl/BatchCombinePre0506.wdl:11]: We usually put spaces between items in a list

File: wdl/SelectBatchesAndColumns.wdl [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		String? docker
+	}
+	
+	Boolean select_batches=defined(batches_opt)
+	Array[String] batches=select_first([batches_opt, ["all"]])
+
+	String docker_set = select_first([docker, "ubuntu:18.04"])
+
+	runtime {
+		docker: docker_set
```
Comment (mwalker174) [inline wdl/SelectBatchesAndColumns.wdl:17]: I believe the default is 3.75 GiB RAM which is probably more than you need. I would suggest reducing to 0.9 GiB.

Comment (mwalker174) [reply wdl/SelectBatchesAndColumns.wdl:17]: Also 3 `preemptible_tries` and 1 `max_retries`

**Comment by epiercehoffman** [comment]
### Updates
The script in SelectBatchesAndColumns.wdl was rewritten in python to address PR review comments. Specific changes:
* Formatting improvements
* Exit if batch not found
* Continue with "COLUMN_DNE" output if column not found
* Check that input file is sample sets table TSV from Terra
* Eliminate scatter operation
* Set default runtime attributes (and provide override input)

### Testing
* cohort2 in [test workspace 1](https://app.terra.bio/#workspaces/broad-firecloud-dsde-methods/batch_combine_testing): BatchCombinePreMerge with File and Array[File] types
* cohort3 in test workspace 1: batch doesn't exist in table --> error
* cohort-v2 in [test workspace 2](https://app.terra.bio/#workspaces/broad-firecloud-dsde-methods/batch_combine_testing_2): BatchCombinePre0506 with no batches specified (--> use all batches)
* cohort3 in [test workspace 3](https://app.terra.bio/#workspaces/broad-firecloud-dsde-methods/batch_combine_testing_3): column doesn't exist in table

Note: if some batches have files for a given column while it is empty for other batches, there will be an error. I think this is a reasonable behavior given that all batches should have the same items, and if not, it is easy enough to remove the batch from the input batches list.

**Comment by epiercehoffman** [comment]
Switching to Terra notebook implementation instead of WDL

**Pull request closed without merging**